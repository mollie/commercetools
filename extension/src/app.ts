import express, { Application } from 'express';
import handler from '../expressHandler';
import morganMiddleware from './logger/morganMiddleware';

const app: Application = express();

app.use(morganMiddleware); // Logs from http level

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use(express.json());

app.use('/', handler);

export default app;
