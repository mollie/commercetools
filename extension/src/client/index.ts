import createMollieClient, { MollieClient } from '@mollie/api-client';
import { ClientBuilder, Client, AuthMiddlewareOptions, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}

export function initialiseCommercetoolsClient(): Client {
  const {
    commercetools: { projectKey, clientId, clientSecret, host, authUrl, scopes },
  } = config;

  const authMiddlewareOptions: AuthMiddlewareOptions = {
    host: authUrl,
    projectKey,
    credentials: {
      clientId,
      clientSecret,
    },
    scopes,
    fetch,
  };

  const httpMiddlewareOptions: HttpMiddlewareOptions = {
    host,
    fetch,
  };

  const client: Client = new ClientBuilder()
    .withClientCredentialsFlow(authMiddlewareOptions)
    .withHttpMiddleware(httpMiddlewareOptions)
    .withUserAgentMiddleware()
    .build();

  return client;
}
