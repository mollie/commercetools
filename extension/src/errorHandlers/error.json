{
    "errors": [
        {
            "code": "Forbidden",
            "message": "Forbidden or Unauthorized - Request to Mollie API failed",
            "extensionExtraInfo": {
                "links": {
                    "documentation": {
                        "href": "https://docs.mollie.com/overview/authentication",
                        "type": "text/html"
                    }
                },
                "title": "Unauthorized Request"
            }
        }
    ]
}