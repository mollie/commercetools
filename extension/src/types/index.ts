// This is not exhaustive
// If you use another commercetools error response code, add it to this enum
export enum CTEnumErrors {
  General = 'General',
  InvalidInput = 'InvalidInput',
  InvalidOperation = 'InvalidOperation',
  Unauthorized = 'Unauthorized',
  SyntaxError = 'SyntaxError',
  SemanticError = 'SemanticError',
  ObjectNotFound = 'ObjectNotFound',
}

export type CTUpdatesRequestedResponse = {
  status: number;
  actions?: Action[];
  errors?: CTError[];
};

export type Action = {
  action: string;
  type?: {
    key: string;
  };
  fields?: {
    actionType: string;
    request?: string;
    response?: string;
    createdAt?: string;
  };
  name?: string;
  value?: string;
  key?: string;
  transaction?: CTTransaction;
  transactionId?: string;
  interactionId?: string;
  state?: CTTransactionState;
};

export type CTError = {
  code: CTEnumErrors;
  message: string;
  extensionExtraInfo?: Object;
};

export type CTMoney = {
  type?: 'centPrecision';
  currencyCode: string;
  centAmount: number;
  fractionDigits?: number;
};

export type CTPayment = {
  id: string;
  amountPlanned: CTMoney;
  transactions?: CTTransaction[];
  key?: string;
  custom?: {
    fields?: {
      paymentMethodsRequest?: string;
      paymentMethodsResponse?: string;
      createPayment?: string;
      createCapture?: string;
      cancelPayment?: string;
      createRefund?: string;
    };
  };
  interfaceInteractions?: CTInterfaceInteraction[];
};

export type CTInterfaceInteraction = {
  id: string;
  actionType: string;
  createdAt?: Date;
  request?: string;
  response?: string;
};

export type CTTransaction = {
  id?: string;
  timestamp?: string;
  type: CTTransactionType;
  amount: CTMoney;
  interactionId?: string;
  state?: 'Initial' | 'Pending' | 'Success' | 'Failure';
};

export enum ControllerAction {
  GetPaymentMethods = 'getPaymentMethods',
  CreateOrder = 'createOrder',
  CreateOrderPayment = 'createOrderPayment',
  CreateOrderRefund = 'createOrderRefund',
  CreateCustomRefund = 'createCustomRefund',
  CreateShipment = 'createShipment',
  UpdateShipment = 'updateShipment',
  CancelOrder = 'cancelOrder',
  NoAction = 'noAction',
  Error = 'error',
}

export enum CTTransactionType {
  Authorization = 'Authorization',
  CancelAuthorization = 'CancelAuthorization',
  Charge = 'Charge',
  Refund = 'Refund',
  Chargeback = 'Chargeback',
}

export enum CTTransactionState {
  Initial = 'Initial',
  Pending = 'Pending',
  Success = 'Success',
  Failure = 'Failure',
}
