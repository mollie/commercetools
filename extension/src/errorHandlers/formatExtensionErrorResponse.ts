import { CTEnumErrors, CTError, CTUpdatesRequestedResponse } from '../types';

/**
 *
 * @param error Either a NodeJS Error or an object containing message, name, field and code
 */
export const formatExtensionErrorResponse = (error: any): CTUpdatesRequestedResponse => {
  const { message, name, field, code } = error;
  const errorMessage = message ? message : 'Error, see logs for more details';
  let ctCode;
  switch (code) {
    case 400:
      ctCode = CTEnumErrors.InvalidInput;
      break;
    case 404:
      ctCode = CTEnumErrors.ObjectNotFound;
      break;
    default:
      ctCode = CTEnumErrors.General;
  }
  const formattedError: CTError = {
    code: ctCode,
    message: errorMessage,
  };

  if (field || name) {
    const extraInfo = Object.assign({}, field && { field }, name && { name });
    formattedError['extensionExtraInfo'] = extraInfo;
  }

  return {
    status: 400,
    errors: [formattedError],
  };
};
