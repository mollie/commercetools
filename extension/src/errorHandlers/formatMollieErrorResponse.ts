import { CTError, CTUpdatesRequestedResponse } from '../types';

// This is based on ApiError from Mollie's SDK
const getExtraInfo = (error: any) => {
  return {
    links: error?.links,
    title: error?.title,
    field: error?.field,
  };
};

/**
 *
 * @param error ApiError from Mollie (extended from Node's Error)
 * Formats error from failed request to Mollie into a
 * Commerce Tools "Validation Failed" response.
 * Docs: https://docs.commercetools.com/api/projects/api-extensions#error
 */
export const formatMollieErrorResponse = (error: any): CTUpdatesRequestedResponse => {
  let formattedError = {} as CTError;

  // 401 or 403
  if (error.status === 401 || error.status === 403) {
    formattedError = {
      code: 'Unauthorized',
      message: 'Forbidden or Unauthorized - Request to Mollie API failed',
      extensionExtraInfo: getExtraInfo(error),
    };
  }
  // 400
  if (error.status === 400) {
    formattedError = {
      code: 'Unauthorized',
      message: 'Request formatted incorrectly or missing information',
      extensionExtraInfo: getExtraInfo(error),
    };
  }
  // 5xx
  if (error.status >= 500) {
    formattedError = {
      code: 'General',
      message: 'Server Error. Please see logs for more details',
      extensionExtraInfo: getExtraInfo(error),
    };
  }

  return {
    status: 400,
    errors: [formattedError],
  };
};
