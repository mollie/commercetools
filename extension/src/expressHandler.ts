import {Request, Response} from 'express';
import handleRequest from './requestHandlers/handleRequest';
import {HandleRequestInput, HandleRequestSuccess} from './types';
import actions from "./requestHandlers/actions";


export default async function handler(req: Request, res: Response) {
    const requestInput = new HandleRequestInput(req.path, req.method, req.body);
    const result = await handleRequest(requestInput);
    if (result instanceof HandleRequestSuccess) {
        return res.status(200).send({actions: actions});

    } else {
        return res.status(result.status).send({errors: result.errors});
    }
};
