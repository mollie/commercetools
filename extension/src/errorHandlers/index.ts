import { formatMollieErrorResponse } from './formatMollieErrorResponse';
import { formatExtensionErrorResponse } from './formatExtensionErrorResponse';

const errorHandler = {
  mollie: formatMollieErrorResponse,
  extension: formatExtensionErrorResponse,
};

export default errorHandler;
