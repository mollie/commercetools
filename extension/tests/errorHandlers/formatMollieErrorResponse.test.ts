import { CTUpdatesRequestedResponse, CTError } from '../../src/types';
import { formatMollieErrorResponse } from "../../src/errorHandlers/formatMollieErrorResponse"

describe('formatMollieErrorResponse', () => {
  it("should return Forbidden when error's status is 403 or 4O1", () => {
    const mockError = {
      status: 401,
      title: "Unauthorized Request",
      links: { 
        documentation: 
          { href: "https://docs.mollie.com/overview/authentication", type:"text/html"}
      } 
    }
    const { status, errors } = formatMollieErrorResponse(mockError);
    expect(status).toBe(400);
    expect(errors?.length).toBe(1)
    errors?.forEach((error) => {
      expect(error).toMatchSnapshot();
    })
  });

  it("should return Bad request when error's status is 400", () => {
    const mockError = {
      status: 400,
      title: "Bad Request",
      links: { 
        documentation: 
          { href: "https://docs.mollie.com/another-link", type:"text/html"}
      } 
    }
    const { status, errors } = formatMollieErrorResponse(mockError);
    expect(status).toBe(400);
    expect(errors?.length).toBe(1)
    errors?.forEach((error) => {
      expect(error).toMatchSnapshot();
    })
  });

  it("should return General error when the error's status is greater or equal to 500", () => {
    const mockError = {
      status: 503,
      title: "Server Error",
      field: "",
      links: { 
        documentation: 
          { href: "https://docs.mollie.com/another-general-error-link", type:"text/html"}
      } 
    }
    const { status, errors } = formatMollieErrorResponse(mockError);
    expect(status).toBe(400);
    expect(errors?.length).toBe(1)
    errors?.forEach((error) => {
      expect(error).toMatchSnapshot();
    })
  });
});
