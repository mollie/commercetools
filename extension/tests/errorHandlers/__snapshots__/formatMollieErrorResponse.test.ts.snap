// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatMollieErrorResponse should return Bad request when error's status is 400 1`] = `
Object {
  "code": "SyntaxError",
  "extensionExtraInfo": Object {
    "field": "amount.value",
    "links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/overview/handling-errors",
        "type": "text/html",
      },
    },
    "title": "Bad Request",
  },
  "message": "Request formatted incorrectly or missing information",
}
`;

exports[`formatMollieErrorResponse should return Forbidden when error's status is 403 or 4O1 1`] = `
Object {
  "code": "Unauthorized",
  "extensionExtraInfo": Object {
    "field": undefined,
    "links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/overview/authentication",
        "type": "text/html",
      },
    },
    "title": "Unauthorized Request",
  },
  "message": "Forbidden or Unauthorized - Request to Mollie API failed",
}
`;

exports[`formatMollieErrorResponse should return General error when the error's status is greater or equal to 500 1`] = `
Object {
  "code": "General",
  "extensionExtraInfo": Object {
    "field": "",
    "links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/overview/handling-errors",
        "type": "text/html",
      },
    },
    "title": "Server Error",
  },
  "message": "Server Error. Please see logs for more details",
}
`;
